tests/tests.wake

```diff
@@ -0,0 +1,143 @@
+# Copyright 2019 SiFive, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You should have received a copy of LICENSE.Apache2 along with
+# this software. If not, you may obtain a copy at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+package test_wake
+from wake import _
+
+export topic wakeTestBinary: Unit => Result (Pair String (List Path)) Error
+
+def wakeToTest Unit = match (subscribe wakeTestBinary)
+    fn, Nil = fn Unit
+    Nil = Pass (Pair "{wakePath}/wake" Nil)
+    _ = Fail (makeError "Two wake binaries declared for testing!")
+
+export def runTests cmdline =
+    require Pass filter = match cmdline
+        Nil      = Pass `.*`
+        arg, Nil = stringToRegExp arg | rmapFail (Fail _.makeError)
+        _        = Fail (makeError "Too many arguments to runTests")
+
+    require Pass (Pair wakeBin _) =
+        wakeToTest Unit
+
+    def tests =
+        sources here `${filter}/(pass|fail)\.sh`
+
+    def prepareReport test =
+        def folder = replace `^${quote here}/` "" test.getPathName
+        def category = replace `/.*$` "" folder
+        def testName = replace `^[^/]*/` "" folder | replace `/(pass|fail)\.sh$` ""
+        def result = runTest test
+        Triple category testName result
+
+    def results =
+        map prepareReport tests
+
+    def formatCategory list =
+        require Triple category _ _, _ = list else "impossible"
+        def reportFailed = match _
+            Triple _ name (Fail (Error cause _)) = Some "    - {name}: {cause}"
+            _ = None
+        def failures = mapPartial reportFailed list
+        def text =
+            def failed = len failures
+            def total = len list
+            def pass = total - failed
+            "  - {category}: {str pass}/{str total} tests pass",
+            failures
+        catWith "\n" text
+
+    def report =
+        results
+        | groupBy (_.getTripleFirst <* _.getTripleFirst)
```

Comment:
`isLT` is much easier to read than `<*` (had to look it up)

Reply:
isLT would need an order, so you'd have to combine it with `<=>*`; ie: `(_.getTripleFirst <=>* _.getTripleFirst | isLT)`. I'm not sure that's an improvement.